design/020-full_dependency_graph_with_build_requires.md

```diff
@@ -0,0 +1,57 @@
+2. Compute which packages need to be built, based on the current existing binaries and the ``--build`` policies
+3. For those packages that need to be built, compute their ``build_requires``, let's say that we need to build ``poco/1.9.4`` from sources, and it has a ``build_requires`` that is ``cmake/3.14``.
+4. Expand the graph with the ``build_requires`` and its transitive ``requires``. In this example, the algorithm would add ``poco/1.9.4``-(build_requires)->``cmake/3.14``.
+5. Go to 2, until no more packages are marked as "necessary to build"
+6. Install the binary packages, downloading pre-compiled binaries, finding them already in the cache or building them from sources
+
+
+This approach has some disadvantages:
+
+- Build-requires can never affect the ``package_id``. Even if this is a user's request, it is impossible by its own definition. Unless a package is marked as “necessary to build”, the ``build_requires`` are not computed. But to know if a package needs to be built or not, first it is necessary to compute its ``package_id``. This chicken-and-egg problem makes impossible that build_requires could affect their consumers ``package_ids``
+- It is more complicated to manage and orchestrate projects builds, for example in CI. Unless careful management of ``--build`` policy is done, it is easy that build_requires could be missing in lockfiles for example.
+- It makes the Conan codebase much more complicated, and the treatment of dependencies more heterogeneous, which makes thorough and flexible management of dependencies more difficult. The extra maintenance slows down development.
+
+
+## Proposal
+
+When the dependency graph is computed, the whole dependency graph will be computed. The algorithm could be outlined as:
+
+1. Starting from the initial conanfile, compute the dependencies and transitive dependencies of ``requires`` and ``build_requires``. We could have at this stage a ``conanfile.txt``->``poco/1.9.4``->``zlib/1.2.11``, and also ``poco/1.9.4``-(build-require)->``cmake/3.14``
+2. Compute which packages need to be built, based on the current existing binaries and the the ``--build`` policies
+3. Install the binary packages, downloading pre-compiled binaries, finding them already in the cache or building them from sources
+
+
+## Detailed Design
+
+This approach would solve the above issues. It will have some other potential issues:
+
+- It might be slightly less efficient, as it might need to resolve some extra recipes as ``cmake/3.14`` that wouldn’t be strictly necessary otherwise. But only the recipes will be necessary, downloading and unzipping the potentially heavy binary packages can still be avoided, so the impact on the overall time might be negligible. In practice, it could suppose and average penalty of around 1-2% of the installation of a graph with existing binaries (much less if something is built from sources)
```

Comment:
As we identified earlier the recipe resolution is a bottleneck :( (when there is no cache, as in some CI configurations).

Reply:
Recipe resolution of a large dependency graph can take some time. But in the context of the ``build_requires``, it is not like projects are having hundreds of different build requirements, but just a very few of them (cmake and similar tools), while we are aware of many users having dependency graphs with hundreds of dependencies. So fetching those build-requires should be relatively small compared with the resolution of the whole graph. Do you have tons of different build_requires in the same graph?

On the other hand, it is possible that a simplified dependency resolution could help in some tasks like parallelizing some tasks in it, and maybe speeding it up too.